pkg/agent/controller/endpoint.go

```diff
@@ -0,0 +1,273 @@
+package controller
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+)
+
+func NewEndpointController(kubeClientSet kubernetes.Interface, serviceImportuid types.UID, serviceImportName,
+	clusterId string) (*EndpointController, error) {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	endpointController := &EndpointController{
+		endPointqueue:     queue,
+		serviceImportUID:  serviceImportuid,
+		serviceImportName: serviceImportName,
+		kubeClientSet:     kubeClientSet,
+		clusterID:         clusterId,
+		stopCh:            make(chan struct{}),
+	}
+
+	return endpointController, nil
+}
+
+func (e *EndpointController) Start(stopCh <-chan struct{}, labelSelector fmt.Stringer) error {
+	e.endpointInformer = cache.NewSharedIndexInformer(
```

Comment:
You don't need a shared Informer since you have a single dedicated event handler.

```suggestion
	e.endpointInformer = cache.NewInformer(
```